examples/5_secret_storage.py

```diff
@@ -39,93 +39,130 @@
 
 
-def bytes_to_str(x):
-    return b64encode(x).decode("utf-8")
-
+class CachedBot(Bot):
+    """
+        Custom bot maintains a cache of secrets for
+        SecretKeyKVStoreClients.
+    """
 
-def str_to_bytes(x):
-    return b64decode(x.encode("utf-8"))
+    @property
+    def secret_kvstore(self):
+        if not hasattr(self, "secret"):
+            # secrets = {team: {namespace: secret}}
+            self.secrets: Dict[str, Dict[str, bytes]] = {}
+        return SecretKeyKVStoreClient(self)
 
 
-class RentalMsg(Enum):
-    ADD = "add"
-    REMOVE = "remove"
-    RESERVE = "reserve"
-    UNRESERVE = "unreserve"
-    LOOKUP = "lookup"
-    LIST = "list"
-    HELP = "help"
+class SecretKeyKVStoreClient(KVStoreClient):
+    """
+        A KVStoreClient that hides the entryKeys from Keybase servers.
 
+        It does so by HMACing entryKeys using a per-(team, namespace) secret,
+        and storing the HMAC instead of the plaintext entryKey. The secret
+        is not expected to change.
 
-class SecretKeyKVStoreClient:
-    """
-        Wrapper around KVStoreClient.
-        HMACs entryKeys using the passed in secret, and stores the HMAC instead
-        of the plaintext entryKey.
-
-        This approach to keeping the entryKeys private stores the plaintext entryKey
-        in the JSON entryValue under the key "_key". Listing all keys requires
-        getting each row in the (team, namespace). Also, this approach does not
-        hide memory access patterns.
+        The plaintext entryKey is stored in the JSON entryValue under the key
+        "_key" to enable listing; listing all keys requires getting each row
+        in the (team, namespace). Also, this approach does not hide memory
+        access patterns.
     """
 
     KEY_KEY = "_key"
+    SECRET_KEY = "_secret"
+    SECRET_NUM_BYTES = 32
+
+    def __init__(self, bot):
+        self.secrets = bot.secrets
+        super().__init__(bot)
 
-    def __init__(self, kvstore: KVStoreClient):
-        self.kvstore = kvstore
+    async def load_secret(self, team, namespace) -> bytes:
+        if team not in self.secrets or namespace not in self.secrets[team]:
+            secret = secrets.token_bytes(self.SECRET_NUM_BYTES)
+            try:
+                # we don't expect self.SECRET_KEY's revision > 0
+                await super().put(
+                    team, namespace, self.SECRET_KEY, bytes_to_str(secret), revision=1
+                )
+            except RevisionError:
+                res: keybase1.KVGetResult = await super().get(
+                    team, namespace, self.SECRET_KEY
+                )
+                secret = str_to_bytes(res.entry_value)
+            # update self.secrets (which also updates self.bot.secrets)
+            if team not in self.secrets:
+                self.secrets[team] = {}
+            self.secrets[team][namespace] = secret
+        return self.secrets[team][namespace]
 
     async def hmac_key(self, secret: bytes, key: str) -> str:
```

Comment:
Why is this function async?

Reply:
good catch...missed it in refactoring. but in the next commit i'm moving the load_secret call into hmac_key, so it will be async